md4c/md4c.c

```diff
@@ -3497,6 +3535,74 @@ md_resolve_links(MD_CTX* ctx, const MD_LINE* lines, int n_lines)
     return 0;
 }
 
+static int
```

Comment:
I have some trouble understanding why this big function is needed. Even if it is needed it has some issues I commented in its body.

Generally, I assumed we can still scan only backwards from the `]]` as before. The only novelty would be that when we reach a `|`, we use it as a beginning of the ID from that point (the closing `MD_MARK` would be expanded to hide it from normal text flow), and that's it.

Also note we then do not have to collect `|` marks beforehand for wiki-links: We reach that char by the scanning over the ID anyway, so it would simply the logic quite a lot.

Reply:
> Generally, I assumed we can still scan only backwards from the `]]` as before. The only novelty would be that when we reach a `|`, we use it as a beginning of the ID from that point (the closing `MD_MARK` would be expanded to hide it from normal text flow), and that's it.

The *closing* mark? Don't you mean the opening mark?

> Also note we then do not have to collect `|` marks beforehand for wiki-links: We reach that char by the scanning over the ID anyway, so it would simply the logic quite a lot.

That would be neat.

Reply:
> The closing mark? Don't you mean the opening mark?

`[[text|id]]`... I.e. the closer mark would be expanded back to cover not only `]]` but whole `|id]]`.

Reply:
Aha, then it would be so, if not for the fact that the wiki link structure is the other way around: `[[id|text]]`.

Reply:
Oops. Mea culpa. So I got it whole wrong and we also need scanning forward from the opener, not backward. Whatever. Does not change the principle.

Reply:
> > Also note we then do not have to collect `|` marks beforehand for wiki-links: We reach that char by the scanning over the ID anyway, so it would simply the logic quite a lot.

Can I get the `MD_MARK` for the `|` by way of it's position in the text somehow?

Reply:
The `|` can be escaped inside the wiki link, so that has to be accounted for. Any pointers?

Reply:
> Can I get the `MD_MARK` for the `|` by way of it's position in the text somehow?

There is no infrastructure for it and it would be better to avoid such need. 

Even doing some binary search would be complicated because although the marks are originally ordered by their natural positions, during the resolving they may be played with and that breaks that property (e.g. they may be collapsed to a zero-length, expanded, there are the dummy marks which actually are only containers for other data so their `beg` and `end` members are misused for other purposes etc.

> The `|` can be escaped inside the wiki link, so that has to be accounted for. Any pointers?

I would likely ignore the marks for that purpose here altogether, it would make it just too complicated. I would simply count the number of `\` just before the pipe (ideally on the fly during the scanning we already do). If it is an odd number, it is escaped.

Reply:
Do you have any handy tips on marking the `|` as resolved?

Reply:
The `|` is still appearing in the output, so I assume I must mark it as resolved somehow. Is there a suitable way to do that in this context, or is there something else I should do?

Reply:
Ouch. A weak spot in my consideration. 

I have an idea, but I am not sure whether it would really work (I do not have all the stuff in my head any more), so take the following with a little bit of salt.

`md_collect_marks()` already places a dummy `MD_MARK` after every `[` because if it gets later resolves as a link we need a space for storing some data.

But if I remember correctly, it is not used at all unless the link is eventually resolved (if it happens at all). So technically we could store some data there which could later be retrieved in the order to get the index of the `|` mark in the `ctx->marks[]`.

I.e. something like this:
1. Make `md_collect_marks()` alweas remember the index of last such dummy after the most recent `[`.
2. When `md_collect_marks()` encounters `|` and it constructs its corresponding `MD_MARK`, we could store its index in the remembered dummy mark (if it is not already set because we want only to register the 1st such pipe).
3. When you are resolving the `[[...]]`, you should be able to dig it from there and verify whether it is between the `[[` and `]]` (and not somewhere later).

It could then also be used for the escaped pipe detection because `md_collect_marks()` escapes escaped characters and does not create `MD_MARK` for escaped chars at all.

Reply:
> The `|` is still appearing in the output, so I assume I must mark it as resolved somehow. Is there a suitable way to do that in this context, or is there something else I should do?

Uff? You may hide it be expanding the opening `[[` to cover not just the ID but also the pipe char.

But we likely still need to do something with the mark in order to hide it from a table cell handling. But not resolving (resolving means, the `md_process_inlines()` would make a stop there to consider calling `enter_span()/leave_span()` callback). I would change it to dummy instead.

Reply:
I have just realized the above suggestion wouldn't work. We do not have any guaranty the pipe has to belong to the most recent `[`.

Counter-example is this:
```
[[[id_in_brackets]|text]
```

Reply:
To sum up, the initial attempt at label handling works, but the following has not been considered:

- possible quadratic behavior (although I cant find this)
- leading quote block marks are counted in the wiki link content limit

Also, you think that there should not be a limit on how many lines the label may span, and I agree with this.

For counting the number of leading characters that make up the block quote, I found some interesting code in the beginning of `md_analyze_line`. Do you think it would be suitable for use here?

And, do you still think the wiki link routine should be inlined in `md_resolve_links`, or could it be contained in its own function?

Reply:
Continuing from where I left off, it looks like `total_indent` as used in `md_analyze_line` has exactly the information that is needed here.

Reply:
One solution is to hand over `total_indent` to the `MD_LINE` that is created in `md_add_line_into_current_block`, but that would mean one more member to `MD_LINE`/`MD_LINE_tag`.

Reply:
> For counting the number of leading characters that make up the block quote, I found some interesting code in the beginning of md_analyze_line. Do you think it would be suitable for use here?
> ...
> Continuing from where I left off, it looks like `total_indent` as used in `md_analyze_line` has exactly the information that is needed here.

No way. This would break everything and it would make the complexity problem worse.

1. It is the main part for detecting starts and ends of blocks, determining their type from line prefix/indentation. The problem is it cannot be reasonably used anywhere else because it inputs needs the type of previous line and also it iterates over arbitrary amount of characters and there is no limit on line length. Would we call it since beginning of document until we reach the line we are interested in? I cannot imagine that.

2. The function `md_analyze_line()` is super expensive. It scans over every char in the line.

It's worse then scanning some part of it manually. If we would hypothetically call `md_analyze_line()` just once per a wiki-link, we have a qudrattic behavior if you please big number of wiki-links to a single enormously long line.

3. Also there are special cases where it would fail horribly, e.g. when nested in something what looks like a link or image:
```
![  [[wiki]] ](/img/url)
```
because in those cases, we artificially create vector of `MD_LINE`s for contents of the image and process its contents partially as a standalone paragraph.

> One solution is to hand over `total_indent` to the `MD_LINE`.

1. It would extremely complicate things e.g. for the image nesting as above or for tables, as there we artificially construct `MD_LINE` too for each table cell.

2. This is the most critical structure for memory overhead of MD4C. We construct a vector of blocks and all their `MD_LINE`s per whole document before the inline processing. So 3 integers instead of 2 in this struct is a big change. I certainly won't do anything like that with a light heart.

Reply:
> * possible quadratic behavior (although I cant find this)

I spent some thinking about it.

Maybe it is not there, at least if the iterating over `MD_MARK`s is done very carefully: If the ID length has a limit (say 100 characters), then it also limits the number of potential `MD_MARK` we have to iterate in order to find the pipe: If we reach any (non-dummy) `MD_MARK` which has `MD_MARK::beg > opener.end + 100`, we know there is no such pipe mark.

We must be careful to now allow iteration over arbitrary (unlimited) stuff. But if we can guarantee it has a (reasonable) limit, it should be ok.

(But sorry, I don't have time today to dig in the code today to verify whether it works that way already. Busy with other things.)

Reply:
I didn't mean to make a call to `md_analyze_line`. What I meant was that `total_indent` could be saved in the `MD_LINE_ANALYSIS` struct, and then handed/copied over to the `MD_LINE` struct in `md_add_line_into_current_block`. That way, in `md_resolve_links`, because we have access to the line(s), we would have access to the `total_indent` value.